Google.PowerShell/Common/AuthenticateWithSdkCredentialsExecutor.cs

```diff
@@ -0,0 +1,159 @@
+﻿/*
+Copyright 2016 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Http;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.PowerShell.Common
+{
+    /// <summary>
+    /// OAuth 2.0 credential for accessing protected resources using an access token.
+    /// This class will get the access token from "gcloud auth print-access-token".
+    /// </summary>
+    public class AuthenticateWithSdkCredentialsExecutor : ICredential, IHttpExecuteInterceptor, IHttpUnsuccessfulResponseHandler
+    {
+        /// <summary>
+        /// This is a cache of the last call to "gcloud auth print-access-token".
+        /// If the current user hasn't changed, and the token hasn't expired, we reuse it
+        /// rather than launching gcloud again.
+        /// </summary>
+        private static ActiveUserToken s_token;
+        private static readonly SemaphoreSlim s_tokenLock = new SemaphoreSlim(1);
+
+        /// <summary>
+        /// Returns the authorization code flow. 
+        /// This is used to revoke token (in RevokeTokenAsync)
+        /// and intercept request with the access token (in InterceptAsync).
+        /// </summary>
+        public IAuthorizationCodeFlow Flow { get; } = new GoogleAuthorizationCodeFlow(
```

Comment:
Let's make this property private, since (if I am reading the code correctly) it is not used outside of the class. Also, perhaps it should just be class field? The ` { get; } ` isn't buying us anything. (We'd need to rename the field to be a lowercase `flow` though.